src/jit.c

```diff
@@ -0,0 +1,1557 @@
+}
+
+static void update_branch_imm(struct jit_state *state,
+                              uint32_t offset,
+                              int32_t imm)
+{
+    assert((imm & 3) == 0);
+    uint32_t insn;
+    imm >>= 2;
+    memcpy(&insn, state->buf + offset, sizeof(uint32_t));
+    if ((insn & 0xfe000000U) == 0x54000000U /* Conditional branch immediate. */
+        || (insn & 0x7e000000U) ==
+               0x34000000U) { /* Compare and branch immediate.  */
+        assert((imm >> 19) == INT64_C(-1) || (imm >> 19) == 0);
+        insn |= (imm & 0x7ffff) << 5;
+    } else if ((insn & 0x7c000000U) == 0x14000000U) {
+        /* Unconditional branch immediate.  */
+        assert((imm >> 26) == INT64_C(-1) || (imm >> 26) == 0);
+        insn |= (imm & 0x03ffffffU) << 0;
+    } else {
+        assert(false);
+        insn = BAD_OPCODE;
+    }
+#if defined(__APPLE__) && defined(__aarch64__)
+    pthread_jit_write_protect_np(false);
+#endif
+    memcpy(state->buf + offset, &insn, sizeof(uint32_t));
+#if defined(__APPLE__) && defined(__aarch64__)
+    pthread_jit_write_protect_np(true);
+#endif
+}
+#endif
+
+static inline void emit_jump_target_offset(struct jit_state *state,
+                                           uint32_t jump_loc,
+                                           uint32_t jump_state_offset)
+{
+    struct jump *jump = &state->jumps[state->n_jumps++];
+    jump->offset_loc = jump_loc;
+    jump->target_offset = jump_state_offset;
+}
+
+/* The REX prefix and ModRM byte are emitted.
+ * The MR encoding is utilized when a choice is available. The 'src' is often
+ * used as an opcode extension.
+ */
+static inline void emit_alu32(struct jit_state *state, int op, int src, int dst)
+{
+#if defined(__x86_64__)
+    emit_basic_rex(state, 0, src, dst);
+    emit1(state, op);
+    emit_modrm_reg2reg(state, src, dst);
+#elif defined(__aarch64__)
+    switch (op) {
+    case 1: /* ADD */
+        emit_addsub_register(state, false, AS_ADD, dst, dst, src);
+        break;
+    case 0x29: /* SUB */
+        emit_addsub_register(state, false, AS_SUB, dst, dst, src);
+        break;
+    case 0x31: /* XOR */
+        emit_logical_register(state, false, LOG_EOR, dst, dst, src);
+        break;
+    case 9: /* OR */
+        emit_logical_register(state, false, LOG_ORR, dst, dst, src);
+        break;
+    case 0x21: /* AND */
+        emit_logical_register(state, false, LOG_AND, dst, dst, src);
+        break;
+    case 0xd3:
+        if (src == 4) /* SLL */
+            emit_dataproc_2source(state, false, DP2_LSLV, dst, dst, R8);
+        else if (src == 5) /* SRL */
+            emit_dataproc_2source(state, false, DP2_LSRV, dst, dst, R8);
+        else if (src == 7) /* SRA */
+            emit_dataproc_2source(state, false, DP2_ASRV, dst, dst, R8);
+        break;
+    default:
+        __UNREACHABLE;
+        break;
+    }
+#endif
+}
+
+/* REX prefix, ModRM byte, and 32-bit immediate */
```

Comment:
This comment should only make sense to x64 backend. Move into the function body.